docs/terminals.md

```diff
@@ -0,0 +1,141 @@
+# Terminals and Standard IO
+All processes on Unix (and Unix-like) operating systems have 3 standard file descriptors (fd) passed to them at start, collectively standard-IO (`stdio`):
+
+* `0`: standard-in (`stdin`), the input stream into the process
+* `1`: standard-out (`stdout`), the output stream from the process
+* `2`: standard-error (`stderr`), the error stream from the process
+
+When creating and running a container via `runc`, it is important to take care to structure the stdio the new container's process receives.
+
+## Terminal Modes
+
+`runc` supports two distinct methods for passing stdio to the container's primary process:
+
+* pass-through
+* new terminal
+
+### Pass-Through
+In pass-through mode, the `stdio` passed to the `runc` call will be passed as is to the primary container. This means that if you do the following:
+
+```
+$ echo input | runc run some_container > /tmp/log.out 2>& /tmp/log.err
+```
+
+Then the stdout of `some_container` will appear in the file `/tmp/log.out`, the stderr will appear in `/tmp/log.err`, and the process will receive input of `input`. `runc`'s stdout was redirected to `/tmp/log.out`, and `runc`, in turn, passed that handle to the container itself.
+
+To invoke pass-through mode, configure your container's `config.json` with `terminal: false` (which is the default).
+
+### New Terminal
+If you do _not_ want to pass your current stdio through to the container, tell it to create a new terminal.
+
+When you do so, `runc` will do the following:
+
+1. Create a new pseudo-terminal (`pty`).
+2. Pass this pty's slave to the container's primary process as its stdio.
+3. Send the pty's master to a process to read/write stdio for the container's primary process (details below).
+
+To invoke new terminal mode, configure your container's `config.json` with `terminal: true`.
+
+## Which Mode To Use
+`runc` itself runs in two modes:
+
+* foreground
+* detached
+
+You can use either terminal mode with either runc mode. However, there are considerations that may indicate preference for one mode over another.
+
+### Foreground
+When running in foreground, `runc` and the container you invoke are most like running a regular foreground process. This is the scenario most likely to be useful for running in pass-through terminal mode. As the example above showed:
```

Comment:
> This is the scenario most likely to be useful for running in pass-through terminal mode.

Unfortunately it's not as simple as this. It really depends on how you're using `pass-through`. If you want to set up your own `stdio` then actually detached makes more sense in a lot of cases (unless you don't want to bother with writing the exit-code handling).

Reply:
My inclination would be to leave this part. It doesn't say "only", but rather "most likely". I am coming at this as someone who hadn't messed around with `runc` before, banged my head on the wall to understand, and am trying to save others the headache. Simplifications, along with options to get deeper, are the best way to explain it to people IMHO.

Reply:
Please do. I think I need to see how you rewrite it once it is in.